### PR TITLE
ingest/pipeline: convert ingest pipeline to YAML

### PR DESCRIPTION
**Pull request by axw**
## Motivation/summary

Convert the ingest pipeline definition to YAML so we can document/comment it. We generate JSON from this, and continue to expect JSON in code for backwards compatibility. The translation from YAML to JSON maintains map ordering.

Defining pipelines in YAML is already supported by the integration package spec, so when we're ready we can drop the generated JSON and add the YAML pipeline definition directly to the package.

## How to test these changes

N/A, non-functional change.

## Related issues

We'll be extending the pipeline for https://github.com/elastic/apm-server/issues/4919, and adding comments would be helpful.

### CODE REVIEW

**Comment by axw** [comment]
looks like a flaky test, saving the error output here for posterity:

```
[2021-03-22T07:25:21.541Z] === RUN   TestApprovedMetrics

[2021-03-22T07:25:21.541Z]     metrics_test.go:58:   map[string]interface{}{

[2021-03-22T07:25:21.541Z]           	"events": []interface{}{

[2021-03-22T07:25:21.541Z]           		map[string]interface{}{"@timestamp": string("2017-05-30T18:53:41.364Z"), "agent": map[string]interface{}{"name": string("elastic-node"), "version": string("3.14.0")}, "ecs": map[string]interface{}{"version": string("dynamic")}, "event": map[string]interface{}{"ingested": string("dynamic")}, ...},

[2021-03-22T07:25:21.541Z]           		map[string]interface{}{"@timestamp": string("2017-05-30T18:53:41.366Z"), "agent": map[string]interface{}{"name": string("elastic-node"), "version": string("3.14.0")}, "ecs": map[string]interface{}{"version": string("dynamic")}, "event": map[string]interface{}{"ingested": string("dynamic")}, ...},

[2021-03-22T07:25:21.541Z]         - 		map[string]interface{}{

[2021-03-22T07:25:21.541Z]         - 			"@timestamp":     string("2017-05-30T18:53:41.366Z"),

[2021-03-22T07:25:21.541Z]         - 			"agent":          map[string]interface{}{"name": string("elastic-node"), "version": string("3.14.0")},

[2021-03-22T07:25:21.541Z]         - 			"ecs":            map[string]interface{}{"version": string("dynamic")},

[2021-03-22T07:25:21.541Z]         - 			"event":          map[string]interface{}{"ingested": string("dynamic")},

[2021-03-22T07:25:21.541Z]         - 			"host":           map[string]interface{}{"ip": string("127.0.0.1")},

[2021-03-22T07:25:21.541Z]         - 			"labels":         map[string]interface{}{"tag1": string("one"), "tag2": float64(2)},

[2021-03-22T07:25:21.541Z]         - 			"metricset.name": string("app"),

[2021-03-22T07:25:21.541Z]         - 			"observer": map[string]interface{}{

[2021-03-22T07:25:21.541Z]         - 				"ephemeral_id":  string("dynamic"),

[2021-03-22T07:25:21.541Z]         - 				"hostname":      string("dynamic"),

[2021-03-22T07:25:21.541Z]         - 				"id":            string("dynamic"),

[2021-03-22T07:25:21.541Z]         - 				"type":          string("apm-server"),

[2021-03-22T07:25:21.541Z]         - 				"version":       string("dynamic"),

[2021-03-22T07:25:21.542Z]         - 				"version_major": string("dynamic"),

[2021-03-22T07:25:21.542Z]         - 			},

[2021-03-22T07:25:21.542Z]         - 			"process":   map[string]interface{}{"pid": float64(1234)},

[2021-03-22T07:25:21.542Z]         - 			"processor": map[string]interface{}{"event": string("metric"), "name": string("metric")},

[2021-03-22T07:25:21.542Z]         - 			"service": map[string]interface{}{

[2021-03-22T07:25:21.542Z]         - 				"language": map[string]interface{}{"name": string("ecmascript")},

[2021-03-22T07:25:21.542Z]         - 				"name":     string("1234_service-12a3"),

[2021-03-22T07:25:21.542Z]         - 				"node":     map[string]interface{}{"name": string("node-1")},

[2021-03-22T07:25:21.542Z]         - 			},

[2021-03-22T07:25:21.542Z]         - 			"system": map[string]interface{}{"process": map[string]interface{}{"cgroup": map[string]interface{}{...}}},

[2021-03-22T07:25:21.542Z]         - 			"user": map[string]interface{}{

[2021-03-22T07:25:21.542Z]         - 				"email": string("user@mail.com"),

[2021-03-22T07:25:21.542Z]         - 				"id":    string("axb123hg"),

[2021-03-22T07:25:21.542Z]         - 				"name":  string("logged-in-user"),

[2021-03-22T07:25:21.542Z]         - 			},

[2021-03-22T07:25:21.542Z]         - 		},

[2021-03-22T07:25:21.542Z]           		map[string]interface{}{"@timestamp": string("2017-05-30T18:53:42.281Z"), "agent": map[string]interface{}{"name": string("elastic-node"), "version": string("3.14.0")}, "byte_counter": float64(1), "dotted": map[string]interface{}{"float": map[string]interface{}{"gauge": float64(6.12)}}, ...},

[2021-03-22T07:25:21.542Z]           	},

[2021-03-22T07:25:21.542Z]           }

[2021-03-22T07:25:21.542Z]         

[2021-03-22T07:25:21.542Z]         Run `make update check-approvals` to verify the diff

[2021-03-22T07:25:21.542Z]         

[2021-03-22T07:25:21.542Z]     server.go:192: log file: /var/lib/jenkins/workspace/pm-server_apm-server-mbp_PR-4977/src/github.com/elastic/apm-server/systemtest/logs/TestApprovedMetrics/apm-server
```

**Comment by axw** [comment]
jenkins run the tests please

**Comment by jalvz** [comment]
While at it, what do you think about putting all the pipelines in a single `default.json` file? Right now we need to reference pipeline names inc. version, which means a non-zero chance of errors that could go easily unnoticed. Same if Kibana changes how pipeline names are generated, etc.

If we were to have all the pipelines in one file it would be a bit more ugly but easier to maintain. And since it is YAML, it is going to be ugly anyways... 🙈

**Comment by axw** [comment]
@jalvz you mean have a single `default.json` pipeline in integration package? That would mean not using named pipelines, right? And that would mean having just one ingest pipeline which does everything. I haven't thought about this in depth. It might be okay, but it might also make it more difficult for users to customise their pipeline. Let's take a look when we come around to removing the old way of doing things?

**Comment by jalvz** [comment]
Yes, exactly. 

And as a matter of fact, I think it is on the contrary: it would probably make customization easier because users wouldn't have to deal with names...

I think if we don't do it we should prioritize fleet tests that exercise all the pipelines (or are least check that they are installed) because it they have the wrong name nothing will crash but some things wouldn't work.

File: ingest/pipeline/generate.go [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+	if err != nil {
+		log.Fatal(err)
+	}
+	yamlDecoder := yaml.NewDecoder(fin)
+	if err := yamlDecoder.Decode(&doc); err != nil {
+		log.Fatal(err)
+	}
+
+	// Convert the document structure into the one expected by libbeat.
+	// e.g. cnvert {a: 1, b: 2, ...} to [{id: a, body: 1}, {id: b, body: 2}, ...]
```
Comment (stuartnelson3) [inline ingest/pipeline/generate.go:50]: `/s/cnvert/convert/`

**Review by stuartnelson3 — APPROVED** [review]
other than a missing vowel, looks :ok_hand:

**Comment by axw** [comment]
> And as a matter of fact, I think it is on the contrary: it would probably make customization easier because users wouldn't have to deal with names...

For others' sake (Juan and I discussed offline): I had in mind that users might want to create their own pipeline which selectively runs the named pipelines, or injects processors. But tbh that's probably going to be a pain, and users would need to know the pipeline names dynamically generated by Fleet, so probably not going to happen anyway.

> I think if we don't do it we should prioritize fleet tests that exercise all the pipelines (or are least check that they are installed) because it they have the wrong name nothing will crash but some things wouldn't work.

If the pipeline name is wrong, then indexing will fail. I just tested this by changing one of the references to something invalid.

Having said all that, I'm not opposed to creating a monolithic pipeline -- fewer moving parts. Let's discuss in weekly.

**Pull request merged**